### PR TITLE
[Merged by Bors] - feat(FieldTheory): define `IsConjRoot` and prove its basic properties

### PR DESCRIPTION
**Pull request by jjdishere**
We define the property `IsConjRoot K x y` as two elements having the same minimal polynomial over `K`. We show that it is equivalent to the existence of an algebra equivalence `σ : L ≃ₐ[K] L` such that `y = σ x`. We also show that if `x` is a separable element over `K` and the minimal polynomial of `x` splits in `L`, then `x` is not in the `K` iff there exists a different conjugate root of `x` in `L` over `K`. Most of the properties have a similar version already proved in Mathlib.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/FieldTheory/Minpoly/IsConjRoot.lean [hunk 1]
```diff
@@ -0,0 +1,306 @@
...
+import Mathlib.FieldTheory.Minpoly.Basic
+import Mathlib.FieldTheory.Adjoin
+import Mathlib.FieldTheory.Normal
+
+/-!
+# Conjugate roots
+
+This file specialize to the property of conjugate roots. Given two elements over some ring `K`,
+these two elements are conjuagte roots over some `K` if they have the same minimal polynomial over
+`K`.
```
Comment (jcommelin) [inline Mathlib/FieldTheory/Minpoly/IsConjRoot.lean:15]:
```suggestion
Given two elements `x` and `y` of some `K`-algebra, these two elements are *conjugate roots* over `K`
if they have the same minimal polynomial over `K`.
```

File: Mathlib/FieldTheory/Minpoly/IsConjRoot.lean [hunk 2]
```diff
@@ -0,0 +1,306 @@
...
+variable {R K L S A B: Type*} [CommRing R] [CommRing S] [Ring A] [Ring B] [Field K] [Field L]
+variable [Algebra R S] [Algebra R A] [Algebra R B]
+variable [Algebra K S] [Algebra K L] [Algebra K A] [Algebra L S]
+
+variable (R) in
+/--
+We say that `x'` is a conjugate root of `x` over `K` if the minimal polynomial of `x` is the
+same as the minimal polynomial of `x'`.
+-/
+def IsConjRoot (x x' : A) : Prop := (minpoly R x) = (minpoly R x')
```
Comment (jcommelin) [inline Mathlib/FieldTheory/Minpoly/IsConjRoot.lean:49]:
```suggestion
We say that `y` is a conjugate root of `x` over `K` if the minimal polynomial of `x` is the
same as the minimal polynomial of `y`.
-/
def IsConjRoot (x y : A) : Prop := (minpoly R x) = (minpoly R y)
```
To match the module docstring...

Please also update below.

**Comment by jcommelin** [comment]
For other reviewers: this API is being built as part of a project leading to Krasner's lemma.
Having `IsConjRoot` leads to more natural and readable statements.

File: Mathlib/FieldTheory/Minpoly/IsConjRoot.lean (hunk 1 above)
Comment (jjdishere) [reply Mathlib/FieldTheory/Minpoly/IsConjRoot.lean:15]: Thank you!

File: Mathlib/FieldTheory/Minpoly/IsConjRoot.lean (hunk 2 above)
Comment (jjdishere) [reply Mathlib/FieldTheory/Minpoly/IsConjRoot.lean]: Thank you! Now all of the variables are `x` and `y` instead of `x` and `x'`.

**Review by jcommelin — COMMENTED** [review]
maintainer merge

File: Mathlib/FieldTheory/Minpoly/IsConjRoot.lean [hunk 3]
```diff
@@ -0,0 +1,310 @@
...
+variable {R K L S A B: Type*} [CommRing R] [CommRing S] [Ring A] [Ring B] [Field K] [Field L]
+variable [Algebra R S] [Algebra R A] [Algebra R B]
+variable [Algebra K S] [Algebra K L] [Algebra K A] [Algebra L S]
+
+variable (R) in
+/--
+We say that `y` is a conjugate root of `x` over `K` if the minimal polynomial of `x` is the
+same as the minimal polynomial of `y`.
+-/
+def IsConjRoot (x y : A) : Prop := (minpoly R x) = (minpoly R y)
```
Comment (Vierkantor) [inline Mathlib/FieldTheory/Minpoly/IsConjRoot.lean:48]: These parentheses can be dropped, right?
```suggestion
def IsConjRoot (x y : A) : Prop := minpoly R x = minpoly R y
```

File: Mathlib/FieldTheory/Minpoly/IsConjRoot.lean [hunk 4]
```diff
@@ -0,0 +1,310 @@
...
+  have : x ∈ (X - C r).aroots S := by
+    rw [mem_aroots]
+    simp [X_sub_C_ne_zero, h ▸ minpoly.aeval R x]
+  simpa [aroots_X_sub_C] using this
+
+/--
+If `x` is a conjugate root of some element `algebraMap R S r` in the image of the base ring, then
+`x = algebraMap R S r`.
+-/
+theorem eq_algebraMap {r : K} {x : S} (h : IsConjRoot K x (algebraMap K S r)) :
```
Comment (Vierkantor) [inline Mathlib/FieldTheory/Minpoly/IsConjRoot.lean:233]: Maybe add a `@[simp]`-tagged `iff` version for this, like you did for `isConjRoot_zero_iff_eq_zero`?

File: Mathlib/FieldTheory/Minpoly/IsConjRoot.lean [hunk 5]
```diff
@@ -0,0 +1,310 @@
...
+theorem isConjRoot_zero_iff_eq_zero' [Nontrivial R] {x : S} [NoZeroSMulDivisors R S]
+    (hf : Function.Injective (algebraMap R S)) : IsConjRoot R 0 x ↔ x = 0 :=
+  ⟨fun h => eq_zero' h hf, fun h => h.symm ▸ rfl⟩
+
+/--
+`x` is a conjugate root of `0` if and only if `x = 0`.
+-/
+@[simp]
+theorem isConjRoot_zero_iff_eq_zero {x : S} : IsConjRoot K 0 x ↔ x = 0 :=
+  isConjRoot_zero_iff_eq_zero' (algebraMap K S).injective
```
Comment (Vierkantor) [inline Mathlib/FieldTheory/Minpoly/IsConjRoot.lean:268]: Could you also add the symmetric version as `@[simp]` lemma, stating `IsConjRoot K x 0 ↔ x = 0`?

**Review by Vierkantor — COMMENTED** [review]
Very nice API development! Everything important I could think of seems to be there. Just a few minor additions to round out everything.

Are you planning to go back and use this definition in existing results? If you don't have any plans at the moment, please add a TODO note and GitHub issue so we don't forget about it.

File: Mathlib/FieldTheory/Minpoly/IsConjRoot.lean [hunk 6]
```diff
@@ -0,0 +1,310 @@
...
+If `y` is a conjugate root of `x`, then `x` is also a conjugate root of `y`.
+-/
+theorem symm {x y : A} (h : IsConjRoot R x y) : IsConjRoot R y x := Eq.symm h
+
+/--
+If `y` is a conjugate root of `x` and `z` is a conjugate root of `y`, then `z` is a conjugate
+root of `x`.
+-/
+theorem trans {x y y': A} (h₁ : IsConjRoot R x y) (h₂ : IsConjRoot R y y') :
+    IsConjRoot R x y' := Eq.trans h₁ h₂
```
Comment (alreadydone) [inline Mathlib/FieldTheory/Minpoly/IsConjRoot.lean:72]: [#6718](https://github.com/leanprover-community/mathlib4/pull/6718/files#r1816397133) needs the Setoid so you may as well add it here.

File: Mathlib/FieldTheory/Minpoly/IsConjRoot.lean [hunk 7]
```diff
@@ -0,0 +1,310 @@
...
+-/
+theorem neg {x y : S} (h : IsConjRoot K x y) :
+    IsConjRoot K (-x) (-y) := by
+  rw [isConjRoot_def, minpoly.neg x, minpoly.neg y, h]
+
+end IsConjRoot
+
+open IsConjRoot
+
+variable [IsDomain S]
```
Comment (alreadydone) [inline Mathlib/FieldTheory/Minpoly/IsConjRoot.lean:107]: Looks like this is only used much later. Can you move it down?

File: Mathlib/FieldTheory/Minpoly/IsConjRoot.lean (hunk 3 above)
Comment (jjdishere) [reply Mathlib/FieldTheory/Minpoly/IsConjRoot.lean:48]: Thanks!

File: Mathlib/FieldTheory/Minpoly/IsConjRoot.lean [hunk 8]
```diff
@@ -0,0 +1,310 @@
...
+The definition of conjugate roots.
+-/
+theorem isConjRoot_def {x y : A} : IsConjRoot R x y ↔ minpoly R x = minpoly R y := Iff.rfl
+
+namespace IsConjRoot
+
+/--
+Every element is a conjugate root of itself.
+-/
+theorem refl {x : A} : IsConjRoot R x x := rfl
```
[the remark this answers is not in the document]
Comment (jjdishere) [reply Mathlib/FieldTheory/Minpoly/IsConjRoot.lean:60]: Thank you!

**Comment by jjdishere** [comment]
> Are you planning to go back and use this definition in existing results? If you don't have any plans at the moment, please add a TODO note and GitHub issue so we don't forget about it.

Thanks for the reminder. I think I will not work on the refactor very soon. I will add a TODO and a Github issue right now.

File: Mathlib/FieldTheory/Minpoly/IsConjRoot.lean (hunk 7 above)
Comment (jjdishere) [reply Mathlib/FieldTheory/Minpoly/IsConjRoot.lean]: Moved. Thank you!

File: Mathlib/FieldTheory/Minpoly/IsConjRoot.lean (hunk 4 above)
Comment (jjdishere) [reply Mathlib/FieldTheory/Minpoly/IsConjRoot.lean]: Thank you! Added.

File: Mathlib/FieldTheory/Minpoly/IsConjRoot.lean (hunk 5 above)
Comment (jjdishere) [reply Mathlib/FieldTheory/Minpoly/IsConjRoot.lean]: Thank you! Added.

File: Mathlib/FieldTheory/Minpoly/IsConjRoot.lean (hunk 6 above)
Comment (jjdishere) [reply Mathlib/FieldTheory/Minpoly/IsConjRoot.lean]: Added. Thank you!

File: Mathlib/FieldTheory/Minpoly/IsConjRoot.lean [hunk 9]
```diff
@@ -0,0 +1,312 @@
...
+  exact ⟨AlgEquiv.ofBijective σ (σ.normal_bijective _ _ _), hσ⟩
+
+/--
+Let `L / K` be a normal field extension. For any two elements `x` and `y` in `L`, `y` is a
+conjugate root of `x` if and only if there exists an `K`-automorphism `σ : L ≃ₐ[K] L` such
+that `y = σ x`.
+-/
+theorem isConjRoot_iff_exists_algEquiv [Normal K L] {x y : L} :
+    IsConjRoot K x y ↔ ∃ σ : L ≃ₐ[K] L, σ x = y :=
+  ⟨exists_algEquiv, fun ⟨_, h⟩ => h ▸ isConjRoot_of_algEquiv _ _⟩
```
[the remark this answers is not in the document]
Comment (jjdishere) [reply Mathlib/FieldTheory/Minpoly/IsConjRoot.lean]: Thanks! I'll exchange `x` and `y` and add this `isConjRoot_iff_orbitRel`.

File: Mathlib/FieldTheory/Minpoly/IsConjRoot.lean [hunk 10]
```diff
@@ -0,0 +1,312 @@
...
+* Prove `IsConjRoot.smul`, if `x` and `y` are conjugate roots, then so are `r • x` and `r • y`.
+
+## Tags
+conjugate root, minimal polynomial
+-/
+
+
+open Polynomial minpoly IntermediateField
+
+variable {R K L S A B: Type*} [CommRing R] [CommRing S] [Ring A] [Ring B] [Field K] [Field L]
```
[the remark this answers is not in the document]
Comment (jjdishere) [reply Mathlib/FieldTheory/Minpoly/IsConjRoot.lean:41]: Thanks!

File: Mathlib/FieldTheory/Minpoly/IsConjRoot.lean [hunk 11]
```diff
@@ -0,0 +1,363 @@
...
+
+* `isConjRoot_iff_exists_algEquiv`: Let `L / K` be a normal field extension. For any two elements
+`x` and `y` in `L`, `IsConjRoot K x y` is equivalent to the existence of an algebra equivalence
+`σ : L ≃ₐ[K] L` such that `y = σ x`.
+* `not_mem_iff_exists_ne_and_isConjRoot`: Let `L / K` be a field extension. If `x` is a separable
+element over `K` and the minimal polynomial of `x` splits in `L`, then `x` is not in the `K` iff
+there exists a different conjugate root of `x` in `L` over `K`.
+
+## TODO
+* move `IsConjRoot` to earlier files and refactor the theorems in field theory using `IsConjRoot`.
```
Comment (jjdishere) [inline Mathlib/FieldTheory/Minpoly/IsConjRoot.lean:30]:
```suggestion
* Move `IsConjRoot` to earlier files and refactor the theorems in field theory using `IsConjRoot`.
```

File: Mathlib/FieldTheory/Minpoly/IsConjRoot.lean [hunk 12]
```diff
@@ -0,0 +1,363 @@
...
+
+/--
+If `y` is a conjugate root of `x` and `z` is a conjugate root of `y`, then `z` is a conjugate
+root of `x`.
+-/
+@[trans] theorem trans {x y z: A} (h₁ : IsConjRoot R x y) (h₂ : IsConjRoot R y z) :
+    IsConjRoot R x z := Eq.trans h₁ h₂
+
+variable (R A) in
+def setoid : Setoid A where
```
Comment (jjdishere) [inline Mathlib/FieldTheory/Minpoly/IsConjRoot.lean:77]:
```suggestion
/--
The setoid structure on `A` defined by the equivalence relation of `IsConjRoot R · ·`.
-/
def setoid : Setoid A where
```

**Comment by jcommelin** [comment]
maintainer merge

**Comment by Vierkantor** [comment]
Looks good to me, thanks!

bors merge

**Pull request closed without merging**